### ISSUE TITLE
Add support for LetsEncrypt certificate acquisition and rotation in Kestrel

### ISSUE THREAD

**Issue by Eilon** [aspnet/KestrelHttpServer#2971]
Original issue: https://github.com/aspnet/Home/issues/1190

cc @blowdart

**Comment by cwe1ss** [comment]
Thx for creating this issue! Some feature requests:

* It should be possible to use the Let’s Encrypt staging server for dev purposes
* It should be possible to create wildcard certificates
* The certificate storage should be pluggable and support multi-instance scenarios
* There should be an Azure Key Vault storage provider

**Comment by Eilon** [comment]
BTW one option to consider is making sure that Kestrel has any/all required hooks for this to be community-implemented (and maybe that's already the case).

**Comment by Tratcher** [comment]
The SNI callback added in 2.1 should be all that's needed to plug in a separate implementation. @natemcmaster did your prototype need any additional hooks?

**Comment by natemcmaster** [comment]
No, but I had to write some code that felt ugly in order to set the SNI callback to use a method on a service from DI.

See https://github.com/natemcmaster/LetsEncrypt/blob/6e5d6d96b20ed2302bfe9bfb5afe3b30d0307488/src/McMaster.AspNetCore.LetsEncrypt/Internal/KestrelOptionsSetup.cs

**Comment by Tratcher** [comment]
Yeah, I wouldn't want an end user to have to write that code but it looks reasonable if a LetsEncrypt implementation provided it.

**Comment by davidfowl** [comment]
We could make the selector an interface instead of just being a delegate.

**Comment by Tratcher** [comment]
And auto-resolve it from DI? The selector needs to be uniquely assignable per endpoint.
